subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/plugins/XCTestConventionPlugin.java

```diff
@@ -0,0 +1,164 @@
+        compile.setModuleName(project.getName());
+
+        compile.setObjectFileDir(buildDirectory.dir("test/objs"));
+
+        DefaultNativePlatform currentPlatform = new DefaultNativePlatform("current");
+        compile.setTargetPlatform(currentPlatform);
+
+        // TODO - make this lazy
+        NativeToolChain toolChain = project.getModelRegistry().realize("toolChains", NativeToolChainRegistryInternal.class).getForPlatform(currentPlatform);
+        compile.setToolChain(toolChain);
+
+        // Add a link task
+        LinkExecutable link = tasks.create("linkTest", LinkExecutable.class);
+        // TODO - need to set basename
+        link.source(compile.getObjectFileDirectory().getAsFileTree().matching(new PatternSet().include("**/*.obj", "**/*.o")));
+        link.lib(configurations.getByName(CppBasePlugin.NATIVE_TEST_LINK));
+        link.setLinkerArgs(Lists.newArrayList("-Xlinker", "-bundle", "-F" + frameworkDir.getAbsolutePath(), "-framework", "XCTest", "-Xlinker", "-rpath", "-Xlinker", "@executable_path/../Frameworks", "-Xlinker", "-rpath", "-Xlinker", "@loader_path/../Frameworks"));
+        PlatformToolProvider toolProvider = ((NativeToolChainInternal) toolChain).select(currentPlatform);
+        Provider<RegularFile> exeLocation = buildDirectory.file(toolProvider.getExecutableName("exe/" + project.getName() + "Test"));
+        link.setOutputFile(exeLocation);
+        link.setTargetPlatform(currentPlatform);
+        link.setToolChain(toolChain);
+
+
+        Provider<Directory> testBundleDir = buildDirectory.dir("bundle/" + project.getName() + "Test.xctest");
+        final CreateXcTestBundle testBundle = tasks.create("createXcTestBundle", CreateXcTestBundle.class);
+        testBundle.dependsOn(link);
+        testBundle.setExecutableFile(link.getBinaryFile());
+        testBundle.setInformationFile(project.file("src/test/resources/Info.plist"));
+        testBundle.setOutputDir(testBundleDir);
+        testBundle.onlyIf(new Spec<Task>() {
+            @Override
+            public boolean isSatisfiedBy(Task element) {
+                return testBundle.getExecutableFile().exists();
+            }
+        });
+
+
+        final XcTest test = tasks.create("test", XcTest.class);
+        test.dependsOn(testBundle);
+        test.setBinResultsDir(buildDirectory.dir("results/test/bin"));
+        test.setTestBundleDir(testBundleDir);
+        test.setWorkingDir(buildDirectory.dir("bundle"));
+        test.getReports().getHtml().setDestination(buildDirectory.dir("reports/test").get());
+        test.getReports().getJunitXml().setDestination(buildDirectory.dir("reports/test/xml").get());
+        test.onlyIf(new Spec<Task>() {
+            @Override
+            public boolean isSatisfiedBy(Task element) {
+                return test.getTestBundleDir().exists();
+            }
+        });
+    }
+
+    private File getSdkPlatformPath() {
+        if (cachedSdkPlatformDir == null) {
```

Comment:
This should be in a service of some kind, so that we do this once per build instead of once per project.